fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java

```diff
@@ -536,42 +537,46 @@ public static Map<String, Map<Table, Set<String>>> generateMvRefMap(
+                                                     Set<String> mvPotentialRefreshPartitionNames) {
         gatherPotentialRefreshPartitionNames(needRefreshMvPartitionNames, baseChangedPartitionNames,
-                baseToMvNameRef, mvToBaseNameRef);
+                baseToMvNameRef, mvToBaseNameRef, mvPotentialRefreshPartitionNames);
     }
 
     private static void gatherPotentialRefreshPartitionNames(Set<String> needRefreshMvPartitionNames,
                                                              Map<Table, Set<String>> baseChangedPartitionNames,
                                                              Map<Table, Map<String, Set<String>>> baseToMvNameRef,
-                                                             Map<String, Map<Table, Set<String>>> mvToBaseNameRef) {
+                                                             Map<String, Map<Table, Set<String>>> mvToBaseNameRef,
+                                                             Set<String> mvPotentialRefreshPartitionNames) {
         int curNameCount = needRefreshMvPartitionNames.size();
-        Map<Table, Set<String>> newBaseChangedPartitionNames = Maps.newHashMap();
-        Set<String> newNeedRefreshMvPartitionNames = Sets.newHashSet();
         for (String needRefreshMvPartitionName : needRefreshMvPartitionNames) {
+            // baseTable with its partitions by mv's partition
             Map<Table, Set<String>> baseNames = mvToBaseNameRef.get(needRefreshMvPartitionName);
-            baseNames.forEach((key, value) -> {
-                newBaseChangedPartitionNames.merge(key, value, (set1, set2) -> {
-                    //we should copy set1 here, otherwise the mvToBaseNameRef will be changed
-                    HashSet<String> set = Sets.newHashSet(set1);
-                    set.addAll(set2);
-                    return set;
-                });
-                for (String baseName : value) {
-                    newNeedRefreshMvPartitionNames.addAll(baseToMvNameRef.get(key).get(baseName));
+            Set<String> mvNeedRefreshPartitions = Sets.newHashSet();
+            for (Map.Entry<Table, Set<String>> entry : baseNames.entrySet()) {
+                Table baseTable = entry.getKey();
+                Set<String> baseTablePartitions = entry.getValue();
+                // base table partition with associated mv's partitions
+                Map<String, Set<String>> baseTableToMVPartitionsMap = baseToMvNameRef.get(baseTable);
+                for (String baseTablePartition : baseTablePartitions) {
+                    // find base table partition associated mv partition names
+                    Set<String> mvAssociatedPartitions = baseTableToMVPartitionsMap.get(baseTablePartition);
+                    mvNeedRefreshPartitions.addAll(mvAssociatedPartitions);
+                }
+
+                Preconditions.checkState(mvNeedRefreshPartitions.contains(needRefreshMvPartitionName));
+                if (mvNeedRefreshPartitions.size() > 1) {
```

Comment:
What if it's 1 ?

Reply:
For one-to-one/many-to-one, base table to mv's mapping will only generate one mv refresh partition.

For one-to-many/many-to-many, it will generate more than one mv refresh partitions.
```
    /**
     * According base table and materialized view's partition range, we can define those mappings from base to mv:
     * <p>
     * One-to-One
     *  src:     |----|    |----|
     *  dst:     |----|    |----|
     * eg: base table is partitioned by one day, and mv is partition by one day
     * </p>
     * <p>
     * Many-to-One
     *  src:     |----|    |----|     |----|    |----|
     *  dst:     |--------------|     |--------------|
     * eg: base table is partitioned by one day, and mv is partition by date_trunc('month', dt)
     * <p>
     * One/Many-to-Many
     *  src:     |----| |----| |----| |----| |----| |----|
     *  dst:     |--------------| |--------------| |--------------|
     * eg: base table is partitioned by three days, and mv is partition by date_trunc('month', dt)
     *  </p>
     *
     *  For one-to-one or many-to-one we can trigger to refresh by materialized view's partition, but for many-to-many
     *  we need also consider affected materialized view partitions also.
     *
     *  eg:
     *  ref table's partitions:
     *   p0:   [2023-07-27, 2023-07-30)
     *   p1:   [2023-07-30, 2023-08-02)
     *   p2:   [2023-08-02, 2023-08-05)
     *  materialized view's partition:
     *   p0:   [2023-07-01, 2023-08-01)
     *   p1:   [2023-08-01, 2023-09-01)
     *   p2:   [2023-09-01, 2023-10-01)
     *
     *  So ref table's p1 has been changed, materialized view to refresh partition: p0, p1. And when we refresh p0,p1
     *  we also need to consider other ref table partitions(p0); otherwise, the mv's final result will lose data.
     */
```